libs/sync/files_state.go

```diff
@@ -0,0 +1,103 @@
+package sync
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/databricks/cli/libs/fileset"
+)
+
+// FilesState keeps track of files on the local filesystem and their corresponding
+// entries in WSFS.
+type FilesState struct {
+	// Map of local file names to their last recorded modified time. Files found
+	// to have a newer mtime have their content synced to their remote version.
+	LastModifiedTimes map[string]time.Time `json:"last_modified_times"`
+
+	// Map of local file names to corresponding remote names.
+	// For example: A notebook named "foo.py" locally would be stored as "foo"
+	// in WSFS, and the entry would be: {"foo.py": "foo"}
+	LocalToRemoteNames map[string]string `json:"local_to_remote_names"`
+
+	// Inverse of LocalToRemoteNames. Together they form a 1:1 mapping where all
+	// the remote names and local names are unique.
+	RemoteToLocalNames map[string]string `json:"remote_to_local_names"`
+}
+
+// Convert an array of files on the local file system to a FilesState representation.
+func toFilesState(ctx context.Context, localFiles []fileset.File) (*FilesState, error) {
+	fs := &FilesState{
+		LastModifiedTimes:  make(map[string]time.Time),
+		LocalToRemoteNames: make(map[string]string),
+		RemoteToLocalNames: make(map[string]string),
+	}
+
+	// Expect no files to have a duplicate entry in the input array.
+	seen := make(map[string]fileset.File)
+	for _, f := range localFiles {
+		if _, ok := seen[f.Relative]; !ok {
+			seen[f.Relative] = f
+		} else {
+			return nil, fmt.Errorf("expected only one entry per file. Found duplicate entries for file: %s", f.Relative)
+		}
+	}
```

Comment:
Did we have this check before as well?

Reply:
No, we did not.

I am aware of the risk here but I feel it's better to not operate blind. My plan is to revert this / fix the root cause if this check causes issues for customers.